File: SearchIn2DMatrix-2.java
//tc=O(m+n)
//sc=O(1)
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix == null || matrix.length == 0) {
            return false;
        }
        int m = matrix.length;
        int n = matrix[0].length;

        int row = m - 1;
        int col = 0;

        while (row >= 0 && col < n) {
            if (matrix[row][col] == target) {
                return true;
            }
            else if (target > matrix[row][col]) {
                col++;
            }
            else {
                row--;
            }
        }
        return false;
    }
}

File: searchIn2DMatrix_2.py
from typing import List
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return False

        height = len(matrix) -1
        width = len(matrix[0]) -1
        
        col = 0
        row = height

        while col <= width and row >= 0:
            if matrix[row][col] == target :
                return True
            elif matrix[row][col] > target :
                row -= 1
            else:
                col += 1
        
        return False